demos/https/iot_demo_https_s3_download_async.c

```diff
@@ -0,0 +1,804 @@
+ */
+static _asyncDownloadData_t _pDownloadDatas[ IOT_HTTPS_DEMO_MAX_ASYNC_REQUESTS ] = { 0 };
+
+/**
+ * @brief Keeps track of the total number of bytes downloaded from the file so far. '
+ *
+ * This is updated in different callback contexts.
+ */
+volatile static uint32_t _bytesFileDownloadedSoFar = 0;
+
+/**
+ * @brief The total size of the file to be downloaded.
+ */
+static uint32_t _fileSize = 0;
+
+/*-----------------------------------------------------------*/
+
+/* Declaration of the demo function. */
+int RunHttpsAsyncDownloadDemo( bool awsIotMqttMode,
+                               const char * pIdentifier,
+                               void * pNetworkServerInfo,
+                               void * pNetworkCredentialInfo,
+                               const IotNetworkInterface_t * pNetworkInterface );
+
+/*-----------------------------------------------------------*/
+
+/**
+ * @brief Get an index to a free HTTPS Client library request resource.
+ *
+ * The resource pool is defined above in the static variables section.
+ *
+ * @return The free index if one is found. -1 if one is not found.
+ */
+static int _getFreeRequestIndex( void )
+{
+    int i = 0;
+    int freeIndex = -1;
+
+    IotMutex_Lock( &( _inUseRequestsMutex ) );
+
+    for( i = 0; i < IOT_HTTPS_DEMO_MAX_ASYNC_REQUESTS; i++ )
+    {
+        if( !_pInUseRequests[ i ] )
+        {
+            _pInUseRequests[ i ] = true;
+            freeIndex = i;
+            break;
+        }
+    }
+
+    IotMutex_Unlock( &( _inUseRequestsMutex ) );
+
+    /* If we exited the loop, then none were found. */
+    return freeIndex;
+}
+
+/*-----------------------------------------------------------*/
+
+/**
+ * @brief Callback for an asynchronous request to append the Range header to the request header buffer.
+ *
+ * @param[in] pPrivData - User private data configured with the HTTPS Client library request configuration.
+ * @param[in] reqHandle - Identifier for the current request in progress.
+ */
+static void _appendHeaderCallback( void * pPrivData,
+                                   IotHttpsRequestHandle_t reqHandle )
+{
+    /* The range value string, of the increment of file to download, from the user private data will be referenced with
+     * this variable. */
+    char * rangeValueStr = ( ( _asyncDownloadData_t * ) ( pPrivData ) )->rangeValueStr;
+
+    IotLogInfo( "Inside of the append header callback for part %s", rangeValueStr );
+    /* The length of the range value string. */
+    uint32_t rangeValueLen = strlen( rangeValueStr );
+    /* Set the header for a range request and check the HTTPS Client library return code. */
+    IotHttpsReturnCode_t status = IotHttpsClient_AddHeader( reqHandle, RANGE_HEADER_FIELD, RANGE_HEADER_FIELD_LENGTH, rangeValueStr, rangeValueLen );
+
+    if( status != IOT_HTTPS_OK )
+    {
+        IotLogError( "Failed to write the header Range: %.*s into the request. With error code: %d", rangeValueLen, rangeValueStr, status );
```

Comment:
If this fails, and the async caller doesn't know it, could the caller end up getting back a different amount of format of data than what they were expecting?

Reply:
I will add cancellation for failures.